src/cmd/s3checker/pkg/s3/s3.go

```diff
@@ -0,0 +1,210 @@
+package s3
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/alcionai/clues"
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+)
+
+const noLockCode = "NoSuchObjectLockConfiguration"
+
+var ErrNoRetention = clues.New("missing ObjectLock configuration")
+
+type Options struct {
+	BucketName string
+	Prefix     string
+
+	Endpoint       string
+	DoNotUseTLS    bool
+	DoNotVerifyTLS bool
+
+	AccessKeyID     string
+	SecretAccessKey string
+	SessionToken    string
+
+	Region string
+}
+
+type ObjInfo struct {
+	Key     string
+	Version string
+}
+
+type Client struct {
+	client     *minio.Client
+	bucketName string
+	prefix     string
+}
+
+// Below init code scavenged from
+// https://github.com/kopia/kopia/blob/83b88d8bbf60f4eb17b40dea440c08b594b9e5d3/repo/blob/s3/s3_storage.go
+
+func getCustomTransport(opt *Options) *http.Transport {
+	if opt.DoNotVerifyTLS {
+		return &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	return http.DefaultTransport.(*http.Transport).Clone()
+}
+
+func New(opt *Options) (*Client, error) {
+	creds := credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&credentials.Static{
+				Value: credentials.Value{
+					AccessKeyID:     opt.AccessKeyID,
+					SecretAccessKey: opt.SecretAccessKey,
+					SessionToken:    opt.SessionToken,
+					SignerType:      credentials.SignatureV4,
+				},
+			},
+			&credentials.EnvAWS{},
+			&credentials.IAM{
+				Client: &http.Client{
+					Transport: http.DefaultTransport,
+				},
+			},
+		})
+
+	if len(opt.BucketName) == 0 {
+		return nil, clues.New("empty bucket name")
+	}
+
+	minioOpts := &minio.Options{
+		Creds:  creds,
+		Secure: !opt.DoNotUseTLS,
+		Region: opt.Region,
+	}
+
+	minioOpts.Transport = getCustomTransport(opt)
+
+	cli, err := minio.New(opt.Endpoint, minioOpts)
+	if err != nil {
+		return nil, clues.Wrap(err, "creating client")
+	}
+
+	return &Client{
+		client:     cli,
+		bucketName: opt.BucketName,
+		prefix:     opt.Prefix,
+	}, nil
+}
+
+func maybeAddObj(prefix string, obj ObjInfo, m map[string]ObjInfo) {
```

Comment:
Does this make any difference even if we add?

Reply:
hmmm well I suppose as long as we don't start keeping a list of items instead of a single item it doesn't really matter. It'll just change which item in the set we pick (first seen vs. random element). The main thing we want is to just select a single item for each prefix so we don't make a bunch of get retention calls later